src/Stratis.Bitcoin/P2P/PeerConnector.cs

```diff
@@ -0,0 +1,219 @@
+            this.network = network;
+            this.ParentParameters = parameters;
+            this.Requirements = nodeRequirements;
+
+            this.currentParameters = this.ParentParameters.Clone();
+            this.currentParameters.TemplateBehaviors.Add(new NodeGroupBehavior(this));
+            this.currentParameters.ConnectCancellation = this.disconnect.Token;
+        }
+
+        /// <summary> The async loop we need to wait upon before we can shut down this feature.</summary>
+        private IAsyncLoop asyncLoop;
+
+        /// <summary> Factory for creating background async loop tasks.</summary>
+        private readonly IAsyncLoopFactory asyncLoopFactory;
+
+        private int busyConnectingToPeers;
+        internal NodesCollection ConnectedNodes { get; private set; }
+
+        /// <summary> The cloned parameters used to connect to peers. </summary>
+        private NodeConnectionParameters currentParameters;
+
+        private CancellationTokenSource disconnect;
+
+        /// <summary> How to calculate a group of an IP, by default using NBitcoin.IpExtensions.GetGroup.</summary>
+        private Func<IPEndPoint, byte[]> groupSelector;
+
+        internal int MaximumNodeConnections { get; set; }
+        private Network network;
+        internal NodeConnectionParameters ParentParameters { get; private set; }
+        internal RelatedNodeGroups RelatedGroups { get; set; }
+        internal NodeRequirement Requirements { get; private set; }
+
+        internal void AddNode(Node node)
+        {
+            Guard.NotNull(node, nameof(node));
+
+            this.ConnectedNodes.Add(node);
+        }
+
+        internal void ConnectToPeersAsync()
+        {
+            if (Interlocked.CompareExchange(ref this.busyConnectingToPeers, 1, 0) == 1)
+                return;
+
+            this.asyncLoop = this.asyncLoopFactory.Run(nameof(this.ConnectToPeersAsync), token =>
+            {
+                ConnectToPeers();
+                return Task.CompletedTask;
+            },
+            this.disconnect.Token,
+            repeatEvery: TimeSpans.RunOnce);
+        }
+
+        private void ConnectToPeers()
+        {
+            while (!this.disconnect.IsCancellationRequested)
+            {
+                if (this.ConnectedNodes.Count >= this.MaximumNodeConnections)
+                {
+                    Thread.Sleep(TimeSpan.FromSeconds(1));
+                    continue;
+                }
+
+                Node node = null;
+
+                try
+                {
+                    var peer = FindPeerToConnectTo();
+                    if (peer == null)
+                        continue;
+
+                    using (var timeoutTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.disconnect.Token))
+                    {
+                        timeoutTokenSource.CancelAfter(5000);
+
+                        this.ParentParameters.PeerAddressManager().PeerAttempted(peer.Endpoint, DateTimeOffset.Now);
```

Comment:
As PeerAddressManager is now a singleton we can pass it in the constructor to PeerConnector.

Also lets use IDateTimeProvider.UtcNow

Reply:
Fixed